cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/profile/ProfileDataset.java

```diff
@@ -29,29 +30,41 @@
 import co.cask.cdap.data2.dataset2.DatasetFramework;
 import co.cask.cdap.data2.dataset2.lib.table.MDSKey;
 import co.cask.cdap.data2.dataset2.lib.table.MetadataStoreDataset;
+import co.cask.cdap.proto.codec.EntityIdTypeAdapter;
 import co.cask.cdap.proto.id.DatasetId;
+import co.cask.cdap.proto.id.EntityId;
 import co.cask.cdap.proto.id.NamespaceId;
 import co.cask.cdap.proto.id.ProfileId;
 import co.cask.cdap.proto.profile.Profile;
 import co.cask.cdap.runtime.spi.profile.ProfileStatus;
+import com.google.common.annotations.VisibleForTesting;
 import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
 
 import java.io.IOException;
 import java.util.ArrayList;
 import java.util.Collections;
+import java.util.HashSet;
 import java.util.List;
+import java.util.Set;
 import javax.annotation.Nullable;
 
 /**
  * Dataset for profile information. This dataset does not wrap its operations in a transaction. It is up to the caller
  * to decide what operations belong in a transaction.
+ * This dataset uses a underlying MetadataStoreDataset, which uses its default column to store the Profile.
+ * It also uses INDEX_COLUMN to store the entity association with the profile.
```

Comment:
the dataset is starting to get more complex. Can you add some javadocs around what the rowkey looks like and what data is stored in what columns

Reply:
Added javadoc